src/action/platform_info.rs

```diff
@@ -0,0 +1,274 @@
+// Copyright 2020 Google LLC
+//
+// Use of this source code is governed by an MIT-style license that can be found
+// in the LICENSE file or at https://opensource.org/licenses/MIT.
+
+extern crate sys_info;
+
+/// Using for `uname` syscall.
+extern crate libc;
+
+use sys_info::{linux_os_release, os_type, hostname};
+use libc::{uname, utsname, c_char};
+use crate::session::{self, Session};
+use std::ffi::CStr;
+use std::option::Option;
+use std::fmt::{Display, Formatter};
+
+use rrg_proto::Uname;
+
+#[derive(Debug)]
+enum Error {
+    CannotGetOSType(sys_info::Error),
+    CannotGetLinuxRelease(sys_info::Error)
+}
+
+impl std::error::Error for Error {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        use Error::*;
+
+        match *self {
+            CannotGetOSType(ref error) => Some(error),
+            CannotGetLinuxRelease(ref error) => Some(error)
+        }
+    }
+}
+
+impl Display for Error {
+    fn fmt(&self, fmt: &mut Formatter) -> std::fmt::Result {
+        use Error::*;
+
+        match *self {
+            CannotGetOSType(ref error) => {
+                write!(fmt, "can't get OS type error: {}", error)
+            },
+
+            CannotGetLinuxRelease(ref error) => {
+                write!(fmt, "can't get Linux release info error: {}", error)
+            }
+        }
+    }
+}
+
+impl From<Error> for session::Error {
+    fn from(error: Error) -> session::Error {
+        session::Error::action(error)
+    }
+}
+
+/// Structure that contains information about client platform.
+#[derive(Default)]
+pub struct PlatformInfo {
+    system: Option<String>,
+    release_name: Option<String>,
+    version_id: Option<String>,
+    machine: Option<String>,
+    kernel_release: Option<String>,
+    fqdn: Option<String>,
+    architecture: Option<String>,
+    node: Option<String>
+}
+
+/// A Response type for `GetPlatformInfo` action
+pub struct Response {
+    /// Client platform information
+    platform_information: PlatformInfo,
+}
+
+/// Funcion that converts raw C-strings into `String` type
+#[inline(always)]
```

Comment:
Any reason for forcing inlinining over letting the compiler decide (or a slightly less "demanding" variant: just `#[inline]`)?

Reply:
I thought that it would be nice if the conversion code was inlined, but probably this solution should be left to the compiler. Thanks!